cpu/nrf52/radio/nrf802154/nrf802154_radio.c

```diff
@@ -0,0 +1,739 @@
+            dev->cb(dev, IEEE802154_RADIO_CONFIRM_TX_DONE);
+            break;
+        case STATE_RX:
+            if (NRF_RADIO->CRCSTATUS) {
+                bool l2filter_passed = _l2filter(rxbuf+1);
+                bool is_auto_ack_en = ack[1];
+                bool is_ack = rxbuf[1] & IEEE802154_FCF_TYPE_ACK;
+                bool ack_req = rxbuf[1] & IEEE802154_FCF_ACK_REQ;
+
+                /* If radio is in promiscuos mode, indicate packet and
+                 * don't event think of sending an ACK frame :) */
+                if (cfg.promisc) {
+                    dev->cb(dev, IEEE802154_RADIO_INDICATION_RX_DONE);
+                }
+                /* If the L2 filter passes, device if the frame is indicated
+                 * directly or if the driver should send an ACK frame before
+                 * the indication */
+                else if (l2filter_passed) {
+                    if (ack_req && is_auto_ack_en) {
+                        timer_set(NRF802154_TIMER, MAC_TIMER_CHAN_ACK, SIFS);
+                        timer_start(NRF802154_TIMER);
+                        _disable();
+                        _state = STATE_ACK;
+                    }
+                    else {
+                        dev->cb(dev, IEEE802154_RADIO_INDICATION_RX_DONE);
+                    }
+                }
+                /* In case the packet is an ACK and the ACK filter is disabled,
+                 * indicate the frame reception */
+                else if (is_ack && !cfg.ack_filter) {
+                    dev->cb(dev, IEEE802154_RADIO_INDICATION_RX_DONE);
+                }
+                /* If all failed, simply drop the frame and continue listening
+                 * to incoming frames */
+                else {
+                    NRF_RADIO->TASKS_START = 1;
+                }
+            }
+            else {
+                NRF_RADIO->TASKS_START = 1;
+            }
+            break;
+        case STATE_ACK:
+            _state = STATE_RX;
+            NRF_RADIO->PACKETPTR = (uint32_t) rxbuf;
+            _disable();
+            /* This will take around 0.5 us */
+            while(NRF_RADIO->STATE != RADIO_STATE_STATE_Disabled);
+            NRF_RADIO->TASKS_RXEN = 1;
+            _set_ifs_timer(false);
+            break;
+        default:
+            assert(false);
+        }
+    }
+
+    cortexm_isr_end();
+}
+
+static int _confirm_on(ieee802154_dev_t *dev)
+{
+    (void) dev;
+    return 0;
+}
+
+static int _request_on(ieee802154_dev_t *dev)
+{
+    (void) dev;
+    nrf802154_on();
+    return 0;
+}
+
+static int _config_phy(ieee802154_dev_t *dev, ieee802154_phy_conf_t *conf)
+{
+    (void) dev;
+    _disable();
+    int8_t pow = conf->pow;
+
+    if (pow < TX_POWER_MIN || pow > TX_POWER_MAX) {
+        return -EINVAL;
+    }
+
+    NRF_RADIO->FREQUENCY = (((uint8_t) conf->channel) - 10) * 5;
+    _set_txpower(pow);
+    return 0;
+}
+
+static int _off(ieee802154_dev_t *dev)
+{
+    /* TODO: implement */
```

Comment:
the problem is, all registers get lost :/. We would need to back up the settings and load them again.

Reply:
This is true for `at86rf215` too.
Maybe this should be handled by the radio HAL, the current configuration is stored there anyway.

Reply:
I think most radios can retain the configuration (at least `at86rf2xx`, `mrf24j40`, `ccXXXX`, etc).

> Maybe this should be handled by the radio HAL, the current configuration is stored there anyway.

The configuration is not stored in the Radio HAL but in the SubMAC. There are valid use cases where we might want to use the Radio HAL without the SubMAC (e.g OpenWSN).

I agree duplicating the PIB (Physical Information Base) is not the way to go, but handling this in the SubMAC is not the best for radios that retain the configuration.

What I think we could so is something like this:
1. Define a structure for the PIB:
```
typedef struct {
    int16_t channel;
    uint8_t page;
    int8_t tx_power;
    ....
} ieee802154_pib_t;
```

2) For radios that don't keep register values on power off, pass a pointer to the `ieee801254_pib_t` while calling the driver's init function.
```
static const ieee802154_pib_t *nrf802154_pib;
int nrf802154_init(iee802154_dev_t *dev, const ieee802154_pib_t *pib)
{
    ....
    nrf802154_pib = pib;
}
```

3) The `request_on` function can read from the `pib` to reset the values.

Note that the `ieee802154_pib_t` can be allocated by the SubMAC descriptor, or next to the driver in case the SubMAC is not used. This should avoid duplication and probably keep the best of both worlds

Reply:
Sounds good.
I never bothered with retaining the config as I would only put the radio to sleep before ending hibernate, where the CPU will do a full reset on init.

Reply:
Hmmmm maybe in such case it would be even better to use a cap to check if the radio retains those registers. If the application doesn't care about storing these values, it can ignore that cap. Otherwise, the upper layer can restore the values anyway.

AFAIS the number of parameters to be restore is not a big thing (around 5 in most cases).

Reply:
Let's say, `IEEE802154_CAP_SLEEP_REGISTER_RETENTION`.

Some radios can be configured to achieve the lowest power mode when calling `off` in compile time. Thus, the caps can be defined like this:
```c
get_caps(ieee802154_dev_t *dev)
{
    ...
#if !AT86RF2XX_USE_LOWEST_LPM
    case IEEE802154_CAP_SLEEP_REGISTER_RETENTION
#endif 
    return true;
default:
    return false;
}
```

Then the upper layer has always information about what to do on power on

Reply:
Here is it: https://github.com/RIOT-OS/RIOT/pull/14997

What do you think?

Reply:
see https://github.com/RIOT-OS/RIOT/pull/14802#issuecomment-690253214